Userland/Applications/Terminal/main.cpp

```diff
@@ -334,9 +335,44 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     TRY(file_menu->try_add_action(open_settings_action));
     TRY(file_menu->try_add_separator());
-    TRY(file_menu->try_add_action(GUI::CommonActions::make_quit_action([](auto&) {
+
+    auto tty_has_foreground_process = [&] {
+        pid_t fg_pid = tcgetpgrp(ptm_fd);
+        return fg_pid != -1 && fg_pid != shell_pid;
+    };
+
+    auto shell_child_process_count = [&] {
+        Core::DirIterator iterator(String::formatted("/proc/{}/children", shell_pid), Core::DirIterator::Flags::SkipParentAndBaseDir);
+        int background_process_count = 0;
+        while (iterator.has_next()) {
+            ++background_process_count;
+            (void)iterator.next_path();
+        }
+        return background_process_count;
+    };
+
+    auto check_terminal_quit = [&]() -> int {
+        bool has_foreground_process = tty_has_foreground_process();
+        int child_process_count = 0;
+        auto get_close_message = [&]() -> String {
+            if (has_foreground_process) {
+                return "There is still a process running in this terminal. Closing the terminal will kill it.";
+            } else if (child_process_count > 1) {
+                return String::formatted("There are background {} processes running in this terminal. Closing the terminal will kill them.", child_process_count);
+            } else {
+                VERIFY(child_process_count == 1);
+                return "There is a background process running in this terminal. Closing the terminal kill it.";
+            }
+        };
+        if (has_foreground_process || (child_process_count = shell_child_process_count()) > 0)
+            return GUI::MessageBox::show(window, get_close_message(), "Close this terminal?", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::OKCancel);
```

Comment:
"Close this terminal?" is a yes/no kinda question.

Reply:
I'm not sure, I feel like the ok/cancel reads better with the close message (not the title).

\> Closing this terminal will kill...
\> Ok

Reply:
hm...not sure, let's keep it like this for now :shrug: